File: javascript/leetcode/278. First Bad Version.js
/**
 * Definition for isBadVersion()
 *
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
  /**
   * @param {integer} n Total versions
   * @return {integer} The first bad version
   */

  return function(n) {
    let start = 0;

    if (n === 1) return 1;
    while (start < n) {
      let mid = start + parseInt((n - start) / 2);
      let bad = isBadVersion(mid);

      if (bad) n = mid;
      else start = mid + 1;
    }
    return start;
  };
};

File: python/leetcode/278. First Bad Version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):


class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        start = 1

        if n == 1:
            return 1
        while start < n:
            mid = start + int((n - start) / 2)
            bad = isBadVersion(mid)

            if bad:
                n = mid
            else:
                start = mid + 1
        return start
